### PR TITLE
fix(pruning): warn when the free-space pruning threshold is too low to ever run

### LINKED ISSUE #6838: User configured free space pruning threshold too low from estimated required free space.
**Description**

- User configured full pruning threshold of 256GB, but the minimum calculated required space is 268, causing it to never be able to run automatically. Nethermnd should have alerted that this is too low.

```
Mar 17 21:06:36 [226357]: 17 Mar 21:06:36 | Not enough disk space to run full pruning. Required 268 GB. Have 165 GB
```

```
"Pruning": {
    "Mode": "Hybrid",
    "FullPruningTrigger" : "VolumeFreeSpace",
    "FullPruningThresholdMb":256000,
    "FullPruningMaxDegreeOfParallelism": 6,
    "FullPruningCompletionBehavior": "ShutdownOnSuccess"
  }
```

**Steps to Reproduce**
In order to replicate the behavior, please provide a detailed list of steps:

**Actual behavior**
Please describe the outcome that occurred as a result of executing the above steps, and explain why it is considered an issue.

**Expected behavior**
Please provide a clear and concise description of what you expected to happen.

**Screenshots**
If applicable, please include screenshots to help illustrate the problem.

**Desktop (please complete the following information):**
Please provide the following information regarding your setup:
 - Operating System: [e.g. Windows]
 - Version: [e.g. 1.17.0]
 - Installation Method: [e.g. GitHub Release/PPA/Homebrew/Docker]
 - Consensus Client: [e.g. Lodestar v1.4.3]

**Additional context**
Please provide any additional context about the problem.

**Logs**
Please include any relevant logs that may help identify the issue.

### PR DESCRIPTION
**Pull request by Matgothmog**
Fixes #6838

### What

`FullPrunerFactory` now warns at startup when `Pruning.FullPruningTrigger=VolumeFreeSpace` is configured with a `FullPruningThresholdMb` too low for automatic full pruning to ever observe enough free disk space to start. `FullPruner.HaveEnoughDiskSpaceToRun` only proceeds once available free space is at least `ChainSizeThresholdFactor` (130%) of the chain's estimated pruning size, but the `VolumeFreeSpace` trigger only fires once available space drops *below* the configured threshold — so a threshold at or under that 130% requirement leaves no free-space value that satisfies both conditions at once, and automatic pruning silently never runs. The warning names the offending config key (`Pruning.FullPruningThresholdMb`) and states both the configured and the required value; it warns and continues, it never throws.

The check is skipped in three cases:

- `AvailableSpaceCheckEnabled=false` — `FullPruner` doesn't gate on free space at all in that mode, so warning would be a false alarm.
- The trigger isn't `VolumeFreeSpace` (e.g. `Manual`, `StateDbSize`) — the free-space race this warning is about doesn't apply.
- The chain has no pruning-size estimate. `ChainSizes.CreateChainSizeInfo` only supplies one for Mainnet and Sepolia; every other chain (including unrecognised ones) returns `null` and is left alone rather than guessed at.

### Why a warning, not a throw

The stock default trips this. `FullPruningThresholdMb` defaults to `256000`, and Mainnet's date-extrapolated required space is already `331565` (both MB) as of this writing and grows over time — so every Mainnet node running `VolumeFreeSpace` on defaults hits this warning today. That's the fix working as intended: the issue reporter was running exactly `256000`. It's also why this warns instead of throwing `InvalidConfigurationException` the way other `Nethermind.Init` eager-validation checks do (`PruningTrieStateFactory.cs`, `NethermindModule.cs`, `FlatWorldStateModule.cs`) — throwing here would refuse to start every one of those nodes on upgrade over a setting that only disables an opportunistic background job, not a correctness requirement.

### Implementation note

`FullPruner.ChainSizeThresholdFactor` is widened from `private const` to `public const` so the factory reads the same number `FullPruner` itself enforces, instead of duplicating it. That's the entire change to `FullPruner.cs`.

#### What types of changes does your code introduce?

- [x] Bugfix (a non-breaking change that fixes an issue)
- [ ] New feature (a non-breaking change that adds functionality)
- [ ] Breaking change (a change that causes existing functionality not to work as expected)
- [ ] Optimization
- [ ] Refactoring
- [ ] Documentation update
- [ ] Build-related changes
- [ ] Other: _Description_

#### Requires testing

- [x] Yes
- [ ] No

#### If yes, did you write tests?

- [x] Yes
- [ ] No

#### Notes on testing

New `FullPrunerFactoryTests` in `Nethermind.Consensus.Test/FullPruning/`: one parameterized test, 7 `[TestCase]`s — threshold far below the estimated required space, threshold far above it, an unknown chain with no pruning-size estimate, threshold exactly at the boundary, threshold 1 MB above the boundary, `AvailableSpaceCheckEnabled=false` with an otherwise-triggering threshold, and trigger `Manual` with an otherwise-triggering threshold. The boundary cases compute the required space live from `ChainSizes`/`FullPruner.ChainSizeThresholdFactor` rather than hard-coding it, since the estimate is date-extrapolated and drifts daily.

`dotnet test Nethermind.Consensus.Test.csproj -c release -- --filter "FullyQualifiedName~FullPrunerFactoryTests"`: 7 total, 7 succeeded, 0 failed. `dotnet format whitespace --verify-no-changes` clean on the three changed files.

#### Requires documentation update

- [ ] Yes
- [x] No

The `FullPruningThresholdMb` config description is unchanged and still accurate — this only adds a startup warning, not a new setting.

#### Requires explanation in Release Notes

- [x] Yes
- [ ] No

Operators running `Pruning.FullPruningTrigger=VolumeFreeSpace` at or near the stock `FullPruningThresholdMb` default will start seeing a new startup warning naming that config key. It's not a regression — automatic pruning was already never starting for them — but it's a new, visible log line on upgrade and worth calling out so it isn't mistaken for something newly broken.

## Remarks

@kamilchodola's question on the issue — whether the 130% (`ChainSizeThresholdFactor`) requirement itself is too conservative — is untouched by this PR. This change only surfaces the effect of that existing factor to operators; it doesn't revisit the factor's value.

`StateDbSize`-trigger gating (the `PathSizePruningTrigger` branch) isn't covered by the new tests. Its constructor resolves a path under `state/` via `fullPruningDb.GetPath`, and pre-creating that directory for the test would couple the test to `GetPath`'s internal path resolution rather than testing the warning logic itself. Not needed for this fix either way — the warning only fires on the `VolumeFreeSpace` trigger, which is what's tested.

### CODE REVIEW

**Review by asdacap — APPROVED** [review]

**Review by manusw7 — APPROVED** [review]

**Comment by LukaszRozmej** [comment]
@claude review

File: src/Nethermind/Nethermind.Blockchain/FullPruning/FullPruner.cs [hunk 1]
```diff
@@ -199,7 +199,7 @@ await WaitForMainChainChange((e) =>
 
         private bool CanStartNewPruning() => _fullPruningDb.CanStartPruning;
 
-        private const long ChainSizeThresholdFactor = 130;
+        public const long ChainSizeThresholdFactor = 130;
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Blockchain/FullPruning/FullPruner.cs]: With this, `* ChainSizeThresholdFactor / 100` is now written out in three places: here in `HaveEnoughDiskSpaceToRun`, in `FullPrunerFactory.WarnIfThresholdBelowRequiredSpace`, and in the test's `ComputeBoundaryThresholdMb`. Since this is becoming public anyway, how about exposing a single helper instead (e.g. `public static long? RequiredFreeSpace(IChainEstimations)`) and using it in all three? Same surface, and the test can't drift from production.

File: src/Nethermind/Nethermind.Init/FullPrunerFactory.cs [hunk 2]
```diff
@@ -82,9 +83,34 @@ ILogManager logManager
...
+    {
+        if (!pruningConfig.AvailableSpaceCheckEnabled) return;
+
+        long? pruningSizeEstimate = chainEstimations.PruningSize;
+        if (pruningSizeEstimate is null) return;
+
+        long requiredSpace = pruningSizeEstimate.Value * FullPruner.ChainSizeThresholdFactor / 100;
+        if (threshold <= requiredSpace && _logger.IsWarn)
+        {
+            _logger.Warn($"Full pruning threshold {threshold.SizeToString(true)} (={threshold.SizeToString()}) is below the estimated {requiredSpace.SizeToString(true)} (={requiredSpace.SizeToString()}) required to run full pruning; automatic pruning may never find enough free disk space to start. Consider raising Pruning.{nameof(IPruningConfig.FullPruningThresholdMb)}.");
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Init/FullPrunerFactory.cs]: Nit: the condition is `<=` but the message says "is below". "at or below" (or "not above") would match.

File: src/Nethermind/Nethermind.Consensus.Test/FullPruning/FullPrunerFactoryTests.cs [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    [TestCase(1L, BlockchainIds.Mainnet, false, true, FullPruningTrigger.Manual, TestName = "Does not warn when the trigger is Manual, even though the threshold is far below the estimated required space")]
+    public void Warns_only_when_threshold_is_below_estimated_required_space(
+        long thresholdMb,
+        ulong chainId,
+        bool expectWarning,
+        bool availableSpaceCheckEnabled = true,
+        FullPruningTrigger trigger = FullPruningTrigger.VolumeFreeSpace,
+        long boundaryOffsetMb = 0)
+    {
+        // A negative thresholdMb is a marker for "the largest whole-megabyte threshold that does
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Consensus.Test/FullPruning/FullPrunerFactoryTests.cs]: A negative `thresholdMb` meaning "compute the boundary", plus a trailing `boundaryOffsetMb` used by one case, is hard to follow and needs this comment to explain it. A `[TestCaseSource]` yielding `TestCaseData` with the boundary computed when the cases are built would remove both the sentinel and the extra parameter.

File: src/Nethermind/Nethermind.Consensus.Test/FullPruning/FullPrunerFactoryTests.cs [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+                    AvailableSpaceCheckEnabled = availableSpaceCheckEnabled
+                },
+                dbProvider,
+                Substitute.For<IBlockTree>(),
+                new StateBoundaryStore(new MemDb(), new MemDb(), retentionWindowBlocks: null),
+                Substitute.For<INodeStorageFactory>(),
+                Substitute.For<INodeStorage>(),
+                Substitute.For<IProcessExitSource>(),
+                new ChainSpec { ChainId = chainId },
+                new Testably.Abstractions.RealFileSystem(),
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Consensus.Test/FullPruning/FullPrunerFactoryTests.cs]: Nothing on this path creates `baseDbPath`: `DiskFreeSpacePruningTrigger`'s ctor only starts a timer, and `GetDriveInfos` only reads. So the `try/finally` cleanup is dead code, and the test depends on the host's real drives. Testably's `MockFileSystem` would remove both.

**Review by LukaszRozmej — COMMENTED** [review]
The logic checks out: the trigger fires on `available < threshold` and `OnPrune` needs `available >= required`, so `threshold <= required` is exactly the dead zone. The skip conditions match `OnPrune`/`HaveEnoughDiskSpaceToRun`, and warn-not-throw is the right call given the Mainnet default already trips it.

One note on the description: this isn't silent today. Once free space drops below the threshold, `HaveEnoughDiskSpaceToRun` logs `Not enough disk space to run full pruning. Required X GB. Have Y GB` on every trigger tick (every 5 min, `FullPruner.cs:218`). The startup warning still adds value (names the config key, fires before the disk fills), but please correct the "silently never runs" wording, especially since the release note relies on it. The simpler alternative is worth considering: add the threshold and `Pruning.FullPruningThresholdMb` to that existing runtime warning, with no factory change.

Also, `ChainSizes` extrapolates linearly from a July 2024 figure, so the warning is only as good as that estimate. That's another reason to keep it a warning.

File: src/Nethermind/Nethermind.Consensus.Test/FullPruning/FullPrunerFactoryTests.cs [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+using NSubstitute;
+using NUnit.Framework;
+
+namespace Nethermind.Consensus.Test.FullPruning;
+
+public class FullPrunerFactoryTests
+{
+    private const string RequiredSpaceWarningSubstring = "required to run full pruning";
+
+    // #6838: a VolumeFreeSpace threshold below FullPruner's own required-space estimate
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Consensus.Test/FullPruning/FullPrunerFactoryTests.cs]: Placement: `FullPrunerFactory` is in `Nethermind.Init`, and the related tests (`DiskFreeSpacePruningTriggerTests`, `FullPruner` tests) live in `Nethermind.Blockchain.Test/FullPruning`. Consensus.Test is an unexpected place to look. Also, this comment repeats the `<remarks>` on the production method and could go.

File: src/Nethermind/Nethermind.Consensus.Test/FullPruning/FullPrunerFactoryTests.cs (hunk 4 above)
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Consensus.Test/FullPruning/FullPrunerFactoryTests.cs]: Correction: I was wrong here. `GetDriveInfos` creates the pruning directory on the real disk through `new DirectoryInfo(dbPath)` whatever `IFileSystem` is passed, so the cleanup does real work and `MockFileSystem` would not help. In 3cb8bfe07c the test keeps `RealFileSystem` and replaces the `try/finally` with `using TempPath`.

**Review by LukaszRozmej — APPROVED** [review]

**Review by rubo — APPROVED** [review]

**Review by benaadams — APPROVED** [review]
